Ability to add multiple ITAs to an OMIS order during creation

teneightfive opened the pull request:
This adds the ability to add multiple ITAs to an OMIS order.

It's tries to create a generic _add another_ functionality for the form wizard that could be replicated elsewhere.

👉  [DEMO](https://datahub-beta2-pr-436.herokuapp.com/omis/create/dcdabbc9-1781-e411-8955-e4115bead28a) 👈 

## What it looks like

![add-another](https://user-images.githubusercontent.com/3327997/29215683-d9ec08c6-7ea3-11e7-9d64-47fbf703e982.gif)

feedmypixel commented on src/apps/omis/apps/create/controllers/assign-ita.js:
```diff
@@ -12,6 +12,67 @@ class AssignItaController extends FormController {
+
+    if (addKey || removeKey) {
+      if (addKey) {
+        if (!isArray(req.form.values[addKey])) {
```
Could just use [flatten](https://lodash.com/docs/4.17.4#flatten)
```
...
if (addKey) {
  req.form.values[addKey] = flatten([req.form.values[addKey]])
  req.form.values[addKey].push('') 
}
...
```

feedmypixel commented on src/apps/omis/apps/create/controllers/assign-ita.js:
```diff
@@ -12,6 +12,67 @@ class AssignItaController extends FormController {
+        if (!isArray(req.form.values[addKey])) {
+          req.form.values[addKey] = [req.form.values[addKey]]
+        }
+        req.form.values[addKey].push('')
```
feels like this needs a comment, maybe:
`// add empty value to array to show add a new add adviser input`

feedmypixel commented on src/apps/omis/apps/create/controllers/assign-ita.js:
```diff
@@ -12,6 +12,67 @@ class AssignItaController extends FormController {
+        req.form.values[addKey].push('')
+      }
+
+      if (removeKey) {
```
could do:
```
if (removeKey) {
  const [ key, index ] = removeKey.split('::')
  req.form.values[key].splice(index, 1)
}
```

feedmypixel commented on src/apps/omis/apps/create/controllers/assign-ita.js:
```diff
@@ -12,6 +12,67 @@ class AssignItaController extends FormController {
+          if (!isArray(fieldValue)) {
+            values[key] = [fieldValue]
+          } else if (!fieldValue.length) {
+            values[key] = ['']
```
feels like this could do with a comment `// add empty value to array to show add a new add adviser input when no advisers`

feedmypixel commented on src/apps/omis/apps/create/controllers/assign-ita.js:
```diff
@@ -12,6 +12,67 @@ class AssignItaController extends FormController {
+
+        if (fieldOptions.repeatable) {
+          if (!isArray(fieldValue)) {
+            values[key] = [fieldValue]
```
could maybe also use [flatten](https://lodash.com/docs/4.17.4#flatten) here 
```
values[key] = flatten([fieldValue.length ? fieldValue : ''])
```

feedmypixel commented on src/templates/_layouts/form-wizard-step.njk:
```diff
@@ -15,15 +15,54 @@
-              error: errors[key].message
-            })) }}
+            {% if props.repeatable %}
+              {% set fieldValues = values[key] or [''] %}
```
potentially this adding of an empty array element needs to be done in one place and explained why its being done

teneightfive replied:
Yes, good point.

I may try and rework this to actually allow the configuration of whether it should start with an empty one or not. In this case of it being optional it may actually be easier to have none to start with and if the user wants to add some they can.

feedmypixel replied:
If you will always need 1 adviser then you may be able to make your life easier and just always have the ability to add 1 or another one rather than adding configuration to it

teneightfive replied:
That's the problem. OMIS has both cases. During the creation, you don't have to add any _ITAs_ (or what we're calling subscribers). 

Then, during the edit state and whilst you're completing the rest of the order you will need to add advisers from post where you will need at least one.

So in the first case it might be better to not have one there by default but the second case it would make sense to already have one empty on first load.

feedmypixel commented on src/templates/_layouts/form-wizard-step.njk:
```diff
@@ -15,15 +15,54 @@
+        {% for key, fieldOptions in options.fields %}
+          {% set props = {} | assign(fieldOptions, {
+            name: key,
+            label: trans(fieldOptions.label),
```
Not part of this work but can we call this `translate` - the shortened version `trans` IMO looses a lot of meaning when compared to `translate`?

teneightfive replied:
Yep fair point, will pick this up as a separate change after as out of scope for this I feel.

teneightfive replied:
Added #450 to change this. If we merge that one first I'll update this branch to use the new full method name.

teneightfive commented:
>Unless I am missing something it feels like this extra input can just be "hardcoded" and added in the view layer and always be there, then this logic to add it in could go potentially?

@feedmypixel you were right, I've removed the logic from the controller which hopefully simplifies that.

I think I will leave this PR with the single state of there always being an empty one by default for now and if we need to configure that option in the future it will be better to do as a separate change.

tyom commented on src/apps/omis/apps/create/controllers/confirm.js:
```diff
@@ -16,7 +16,9 @@ class ConfirmController extends FormController {
-      values.ita = find(advisers.results, { id: values.ita })
+      values.ita = values.ita.map((ita) => {
+        return find(advisers.results, { id: ita }).name
+      })
```
Do you need to remove `undefined`s afterwards?

teneightfive replied:
Empty and `undefined`s are removed during the [`saveValues`](https://github.com/uktrade/data-hub-frontend/pull/436/files#diff-0df20b7f65681a461e68876f362df584R40) call.

teneightfive replied:
I have just made this slightly more robust though so it doesn't fall over.

tyom commented on src/apps/omis/apps/create/controllers/confirm.js:
```diff
@@ -32,6 +34,10 @@ class ConfirmController extends FormController {
     try {
       const order = await Order.save(req.session.token, data)
 
+      await Order.saveSubscribers(req.session.token, order.id, data.ita.map((id) => {
```
Could you extract third argument into a variable to make it clear that you're passing  request body?
